### DISCUSSION TITLE
How to add one to many polymorphic relationship ?

### DISCUSSION THREAD

**Discussion by samimhamza-pgl** [pocketbase/pocketbase#2087] [Q&A] [answered]
I have four kind of users (applicant, employers, ...) that can login to system and I don't want to create four auth collections. instead I want to have a single users table and four table for each type of my users that login. And I want to have all my authentication in users table.
as I came from Laravel, in Laravel we have polymorphic relations that one foreign key is representing of many tables or collections. for implementing such kind of relation we can add two column loginable_id and loginable_type, which loginable type represent the collection or table that loginable_id came from.
This help me to have all auth login in one collection and all emails and password are saved in single table.
I want to create such kind of relation in pocketbase also can anyone know how to implement this ?

**Comment by ganigeorgiev** [comment, marked as answer]
I'm sorry, I'm not sure that I understand what you are trying to do.

>  I want to have all my authentication in users table.

If you want to keep the authentication in 1 users table, then what is the purpose of the polymorphic relation?

> loginable_id and loginable_type, which loginable type represent the collection or table that loginable_id came from.

Based on this, isn't `loginable_type` be always "users" if you want to keep it in 1 users table?

If all your user types shares the same login options, then you can define a `select` field "role" in the `users` collection. Then in the API rules, you can check the role by `@request.auth.role = "applicant"`.

You can implement a polymorphic relations if you want to and it shouldn't be much different then what you do with Laravel. There are no helper queries for that but each collection is represented as SQLite table under the hood so you can create a custom query when [using PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) or create a "View collection" with a custom `SELECT` statement.

**Reply by samimhamza-pgl** [reply]
thanks for your answer, but let me explain a little more.
polymorphic relationship is a relationship that none of sql databases support. but most of frameworks such as laravel and django has this relationship.
in this relationship we can have a foreign key that can relate to more than one table or collection.
please check once this url for better explanation of polymorphic relationships: https://laravel.com/docs/10.x/eloquent-relationships#one-to-many-polymorphic-relations

**Reply by samimhamza-pgl** [reply]
in regard to loginable_type column, in my case it has four value, applicant, employer, agency and admin, that each of this value is a collection name. and loginable_id is the foreign key of one of these four kind(collection).

**Reply by ganigeorgiev** [reply]
@samimhamza-pgl 

We don't use foreign keys on db level. This is because the relation could be "dynamic" (eg. the collectionId can change programmatically or be "multiple", represented as serialized json array) and relations management is handled on app level.

I'm aware of Laravel's polymorphic relationships support and this is why I mentioned that we don't have built in query helpers for this and you have to handle it on your own, aka. treat the rel pk (id) column as indexed text value and apply the extra "type" conditions to your query manually.

But again, I still don't understand why do you need it and what you are trying to do.

**Reply by samimhamza-pgl** [reply]
I need that because I have one login form in frontend and multiple type of users and each one of them has different fields or columns. so I can't create one collection for all types of users.
but when fetching from users table I want to expand this relation to also fetch data of one of these four types along with the users table.

**Reply by ganigeorgiev** [reply]
If all user types have different fields wouldn't it better to have 4 different auth collections? Wouldn't all other collections have specific references to those types?

You still can use a single login form if you apply a failover flow (aka. try to login first with one collection and if fail try with another, until all are checked) or show a select input where users can choose their type before sending the auth request.

> but when fetching from users table I want to expand this relation to also fetch data of one of these four types along with the users table.

The `?expand` query parameter will not work with polymorphic relations out of the box because it doesn't support conditions (aka. the additional "type" check that you'll have to apply). As mentioned earlier, you can do it with a custom query, but it is more involving.

An alternative approach could be to inverse the relation and use ["Indirect expand"](https://pocketbase.io/docs/expanding-relations/#indirect-expand), aka. structure like:

| users (_auth_) |
| ----------- |
| id |
| email |
| ...   |


| applicants (_base_) |
| ----------- |
| id |
| user (_`relation` to users_) |
| ...   |

| employers (_base_) |
| ----------- |
| id |
| user (_`relation` to users_) |
| ...   |

etc.

`/api/collections/users/records/RECORD_ID?expand=applicants(user)`

The above also can allow a single user to have more than one profile/type.

**Reply by samimhamza-pgl** [reply]
now I understand how to implement it. thank you for your feedback
